### PR TITLE
VPC Network configuration added

### PR DESCRIPTION
**Pull request by VaishnaviGopal**

### CODE REVIEW

File: main.tf [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+  name  = var.vpc
+}
+
+data "ibm_resource_group" "resource_group" {
+  name = var.resource_group != null ? var.resource_group : "Default"
+}
+
+resource "ibm_is_vpc" "vpc" {
+  count                       = var.create_vpc ? 1 : 0
+  name                        = var.vpc_name
```
Comment (umarali-nagoor) [inline main.tf]: Can we use single variable for get VPC name from user for both resource & data_source

File: main.tf [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+  next_hop      = each.value["next_hop"]
+  action        = lookup(each.value, "action", null)
+  destination   = lookup(each.value, "destination", null) != null ? ibm_is_subnet.subnets[each.value["subnet_name"]].ipv4_cidr_block : null
+}
+
+#####################################################
+# Create Network ACL
+#####################################################
+
+locals {
```
Comment (umarali-nagoor) [inline main.tf]: Create a separate file called locals.tf and copy this content there

File: main.tf [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+      destination = r.destination
+      direction   = r.direction
+      icmp        = lookup(r, "icmp", null)
+      tcp         = lookup(r, "tcp", null)
+      udp         = lookup(r, "udp", null)
+    }
+  ]
+}
+
+module "network_acl" {
```
Comment (umarali-nagoor) [inline main.tf]: Why are we having network_acl module block ... can we have the "ibm_is_network_acl" resource directly

File: main.tf [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+  resource_group_id = data.ibm_resource_group.resource_group.id
+  rules             = local.acl_rules
+  tags              = var.acl_tags
+}
+
+#####################################################
+# Create Subnets
+#####################################################
+
+module subnets {
```
Comment (umarali-nagoor) [inline main.tf]: Similar comment, can we have "ibm_is_subnet"  resource directly here

File: main.tf [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+  default_routing_table_name  = (var.default_routing_table_name != null ? var.default_routing_table_name : null)
+  tags                        = (var.vpc_tags != null ? var.vpc_tags : [])
+}
+
+#####################################################
+# Create Routes
+#####################################################
+
+resource "ibm_is_vpc_routing_table" "routing_tables" {
+  for_each                      = { for r in var.routing_table_list : r.name => r }
```
Comment (umarali-nagoor) [inline main.tf]: sorry, where have we defined "routing_table_list" ?

File: .github/workflows/validate_terraform.yml [hunk 6]
```diff
@@ -0,0 +1,39 @@
+name: ci
```
Comment (umarali-nagoor) [inline .github/workflows/validate_terraform.yml:1]: Upon PR raise why no github action is triggered ?

File: .pre-commit-config.yaml [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+# Terraform Fmt      : Used to rewrite Terraform configuration files to a canonical format and style.
+# Terraform Validate : Validates the configuration files in a directory, referring only to the configuration and not accessing any remote services such as remote state, provider APIs, etc
+repos:
+- repo: git://github.com/antonbabenko/pre-commit-terraform
+  rev: v1.45.0
+  hooks:
+    - id: terraform_fmt
+- repo: git://github.com/pre-commit/pre-commit-hooks
+  rev: v3.4.0
+  hooks:
```
Comment (umarali-nagoor) [inline .pre-commit-config.yaml:14]: Add detect secret hook as part of pre-commit checks. Take observability module as reference
https://github.com/terraform-ibm-modules/terraform-ibm-observability/blob/main/.pre-commit-config.yaml

File: examples/README.md [hunk 8]
```diff
@@ -0,0 +1,14 @@
+Use this README.md file to capture a table with trivial use cases such as deleting an already provisioned resource or changing the service plan of a provisioned resource.
```
Comment (umarali-nagoor) [inline examples/README.md]: Make changes to this file according to network module

File: examples/example_1/.gitignore [hunk 9]
```diff
@@ -0,0 +1,37 @@
+# Local .terraform directories
```
Comment (umarali-nagoor) [inline examples/example_1/.gitignore]: please remove the directories like "example_1"

File: examples/example_2/.gitignore [hunk 10]
```diff
@@ -0,0 +1,37 @@
+# Local .terraform directories
```
Comment (umarali-nagoor) [inline examples/example_2/.gitignore]: Please remove the directories like "example_2"

File: module.yaml [hunk 11]
```diff
@@ -0,0 +1,26 @@
+# Sample metadata file
```
Comment (umarali-nagoor) [inline module.yaml]: Update module.yaml file with valid data

File: modules/module_1/README.md [hunk 12]
```diff
@@ -0,0 +1,54 @@
+# <Module_Name_Here> Example
```
Comment (umarali-nagoor) [inline modules/module_1/README.md]: Remove sample directories like "module_1"

File: output.tf [hunk 13]
```diff
@@ -0,0 +1,22 @@
+/********************************************************************
```
Comment (umarali-nagoor) [inline output.tf]: Capture outputs here from root modules for example you are creating a vpc as part of root module ... just output the respective id

File: variables.tf [hunk 14]
```diff
@@ -0,0 +1,212 @@
...
+  type        = string
+  default     = null
+}
+
+
+#####################################################
+# Optional Parameters
+#####################################################
+
+variable "vpc_name" {
```
Comment (umarali-nagoor) [inline variables.tf]: Can VPC name be empty ... as this variable is used to create a new VPC or read an existing VPC

File: variables.tf [hunk 15]
```diff
@@ -0,0 +1,212 @@
...
+  default     = []
+}
+
+variable "enable_gateway" {
+  description = "True to create new Gateway and enable it to subnets"
+  type        = bool
+  default     = false
+}
+
+variable "floating_ip" {
```
Comment (umarali-nagoor) [inline variables.tf:97]: Can we define the content types of map

Comment (VaishnaviGopal) [reply variables.tf:97]: since floating ip can nhave either IP ID or address.. we can't define content type of map.

File: README.md [hunk 16]
```diff
@@ -10,4 +10,103 @@ Scenarios to construct sub-modules
...
+All optional parameters, by default, will be set to `null` in respective example's variable.tf file. You can also override these optional parameters.
+
+## Inputs
+
+| Name                              | Description                                           | Type   | Default | Required |
+|-----------------------------------|-------------------------------------------------------|--------|---------|----------|
+| create\_vpc | True to create new VPC. False if VPC is already existing and subnets, gateways are to be added to it | bool | n/a | yes |
+| vpc\_name | Name of the vpc. Required only if Creating a new VPC | string | n/a | no |
+| resource\_group\_id | ID of the resource group | string | n/a | no |
+| classic\_access | Indicates whether this VPC should be connected to Classic Infrastructure. | bool | false | no |
```
Comment (albee-jhoney) [inline README.md:73]: All the boolean variable must start with `is_<question>`.

File: README.md [hunk 17]
```diff
@@ -10,4 +10,103 @@ Scenarios to construct sub-modules
...
+| default\_address\_prefix | Indicates whether a default address prefix should be automatically created for each zone in this VPC.  | string | auto | no |
+| default\_network\_acl\_name | Name of the Default ACL of the VPC | string | n/a | no |
+| default\_security\_group\_name | Name of the Default Security Group of the VPC | string | n/a | no |
+| default\_routing\_table\_name | Name of the Default Routing Table of the VPC  | string | n/a | no |
+| vpc\_tags | List of tags to attach to the VPC | list(string) | n/a | no |
+| address\_prefixes | List of Prefixes for the vpc | list(object) | n/a | no |
+| locations | zones per region | list(string) | n/a | no |
```
Comment (albee-jhoney) [inline README.md]: Is this `locations` variable used in variables.tf ?

File: README.md [hunk 18]
```diff
@@ -10,4 +10,103 @@ Scenarios to construct sub-modules
...
+To destroy the VPC and all related resources
+```
+terraform destroy -var-file=./input.tfvars
+```
+
+## Note
+
+All optional parameters, by default, will be set to `null` in respective example's variable.tf file. You can also override these optional parameters.
+
+## Inputs
```
Comment (albee-jhoney) [inline README.md:65]: Note, the input variable `names` in the following table & the related `description` **must be the same** in the following location
* Readme.md
* variables.tf
* metadata.yaml

File: README.md [hunk 19]
```diff
@@ -10,4 +10,103 @@ Scenarios to construct sub-modules
...
+| acl\_rules | List of Network ACL Rules that are to be attached to the ACL  | list(object) | n/a | no |
+| acl\_tags | List of tags to attach  | list(string) | n/a | no |
+
+
+## Outputs
+
+| Name | Description |
+|------|-------------|
+| vpc\_id | The ID of the VPC |
+| vpc\_default\_security\_group |The ID of the vpc default security group |
```
Comment (albee-jhoney) [inline README.md:102]: If the variable is returning the ID, it is more readable with ID in the name of the variable itself.  Eg. `vpc_default_security_group_id`
This comment applies for the other variables also.

File: examples/acl/variables.tf [hunk 20]
```diff
@@ -0,0 +1,46 @@
...
+# Optional Parameters
+#####################################################
+
+variable "resource_group" {
+  description = "Resource group name"
+  type        = string
+  default     = null
+}
+
+variable "rules" {
```
Comment (umarali-nagoor) [inline examples/acl/variables.tf:37]: Define the contents of rules variable

variable "rules" {
  description = "List of rules that are to be attached to the Network ACL"
  type = list(object({
    name        = string
    action      = string
    source      = string
    destination = string
    direction   = string
    icmp = object({
      code = number
      type = number
    })
    tcp = object({
      port_max        = number
      port_min        = number
      source_port_max = number
      source_port_min = number
    })
    udp = object({
      port_max        = number
      port_min        = number
      source_port_max = number
      source_port_min = number
    })
  }))
  default = []
}

File: examples/acl/README.md [hunk 21]
```diff
@@ -0,0 +1,32 @@
+# Network ACL Module Example
+
+This module is used to create a Network ACL
```
Comment (umarali-nagoor) [inline examples/acl/README.md:3]: This is redundant as below line giving the purpose of this template

File: test/vpc_test.go [hunk 22]
```diff
@@ -0,0 +1,31 @@
...
+)
+
+func TestAccIBMISVPC(t *testing.T) {
+	t.Parallel()
+
+	// Construct the terraform options with default retryable errors to handle the most common retryable errors in
+	// terraform testing.
+	terraformOptions := terraform.WithDefaultRetryableErrors(t, &terraform.Options{
+		// The path to where our Terraform code is located
+		TerraformDir: ".",
```
Comment (umarali-nagoor) [inline test/vpc_test.go:16]: in current directory we don't have any templates to execute ... I think it should be pointing to example directory

File: metadata.yml [hunk 23]
```diff
@@ -0,0 +1,317 @@
...
+###################################
+root:
+  - path: ""
+    name: "vpc-network"
+    provider_dependencies:
+      - name: "ibm"
+        namespace: "IBM-Cloud"
+        source: "IBM-Cloud/ibm"
+        version: ""
+    inputs:
```
Comment (umarali-nagoor) [inline metadata.yml:25]: Please add other sections as well like outputs, resources, datasource, modules, locals etc

File: README.md [hunk 24]
```diff
@@ -1,13 +1,110 @@
-# terraform-ibm-vpc-network
+#  IBM Cloud terraform-ibm-vpc-network - Terraform Module
 Terraform Module Repository which creates IBM VPC Network infrastructure
 
-Root Module consists of all VPC Network resources.
+This module a collection of modules that make it easier to provision VPC Network resources.
```
Comment (umarali-nagoor) [inline README.md:4]: correct the sentense

**Review by hkantare — DISMISSED** [review]

File: examples/acl/variables.tf (hunk 20 above)
Comment (VaishnaviGopal) [reply examples/acl/variables.tf:37]: adding readme block for the object type should suffice

**Pull request merged**